cirq-core/cirq/ops/control_values.py

```diff
@@ -0,0 +1,453 @@
+            boolean whether the two objects are equivalent or not.
+        """
+
+    def pop(self) -> Optional['ControlValues']:
+        """Removes the last control values combination."""
+        if self._nxt is None:
+            return None
+        return type(self)(self._combinations, self._nxt.pop())
+
+    def link(self, other) -> 'ControlValues':
+        """Links two ControlValues sequences.
+
+        Args:
+            other: An instance of a descendent type of ControlValues
+
+        Returns:
+            A new object that represents the product of the combinations in the two objects.
+        """
+        if self._nxt is None:
+            return type(self)(self._combinations, other)
+        return type(self)(self._combinations, self._nxt.link(other))
+
+    @abc.abstractmethod
+    def identifier(self, companions: Sequence[Union[int, 'cirq.Qid']]):
+        """Attaches the companions to their control values (used in comparisons).
+
+        Args:
+            companions: a sequence of Qid or ints.
+
+        Returns:
+            A sequence of zipped companions to their control values.
+        """
+
+    @abc.abstractmethod
+    def check_dimensionality(
+        self,
+        qid_shape: Optional[Union[Tuple[int, ...], List[int]]] = None,
+        controls: Optional[Union[Tuple['cirq.Qid', ...], List['cirq.Qid']]] = None,
+        offset=0,
+    ) -> None:
+        """Checks the dimentionality of control values with respect to qid_shape or controls.
+            *At least one of qid_shape and controls must be provided.
+            *if both are provided then controls is ignored*
+
+        Args:
+            qid_shape: Sequence of shapes of the control qubits.
+            controls: Sequence of Qids.
+            offset: starting index.
+        Raises:
+            ValueError:
+                - if none of qid_shape or controls are provided or both are empty.
+                - if one of the control values violates the shape of its qubit.
+        """
+
+    @abc.abstractmethod
+    def are_same_value(self, target_value) -> bool:
+        """Checks whether all control values are the same and equal target_value.
+
+        Args:
+            target_value: a value to check whether all control values are eqaul to or not.
+
+        Returns:
+            True/False whether all same control values equal target_value.
+        """
+
+    @abc.abstractmethod
+    def arrangements(self):
+        """Returns a representation of the control values,
+        The representation can be used in comparisons and to rebuild the object."""
+
+    @staticmethod
+    @abc.abstractmethod
+    def factory(val) -> 'ControlValues':
+        """Builds a ControlValue from values."""
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        return {
+            '_combinations': self._combinations,
+            '_nxt': self._nxt._json_dict_() if self._nxt else None,
+        }
+
+
+@ControlValues.register
+class SimpleControlValues(ControlValues):
```

Comment:
Let's bikeshed this during synq. I don't think users will find the name intuitive. Perhaps `PerQubitControlValues`? Also the docstring could use some examples to make it clear.